catx/catx.py

```diff
@@ -59,58 +69,75 @@ def __init__(
 
         self.discretization_parameter = discretization_parameter
         self.bandwidth = bandwidth
-        self.network_builder = network_builder
+        self.custom_network = custom_network
         self.optimizer = optimizer
-        self.rng_key = rng_key
         self.tree_params = TreeParameters.construct(
             bandwidth=bandwidth, discretization_parameter=discretization_parameter
         )
 
         self._action_min = action_min
         self._action_max = action_max
 
-        self.is_initialized = False
+        self._is_initialized = False
 
         self._forward_fn: Wrapped
         self._forward_single_depth_fns: Dict[int, Wrapped]
-        self._params: hk.Params
-        self._depth_params: Dict[int, hk.Params]
-        self._opt_states: Dict[int, optax.OptState]
 
     def sample(
-        self, obs: Observations, epsilon: float
-    ) -> Tuple[Actions, Probabilities]:
+        self,
+        obs: Observations,
+        epsilon: float,
+        state: CATXState,
+    ) -> Tuple[Actions, Probabilities, CATXState]:
         """Samples an action from the tree.
 
         Args:
             obs: the observations, i.e., batched contexts.
             epsilon: probability of selecting a random action.
+            state: holds the CATX's training state.
 
         Returns:
             actions: sampled actions from the tree using epsilon-greedy.
             probabilities: the probability density value of the sampled actions.
+            state: holds the CATX's training state.
         """
 
         obs = jnp.asarray(obs)
 
-        if not self.is_initialized:
-            self._init(obs=obs, epsilon=epsilon)
+        if not self._is_initialized:
+            raise RuntimeError(
+                'CATX instance must initialized before calling "sample" or "learn".'
+            )
```

Comment:
I am not sure this is jittable since it makes the code depend on self which has to be static.

Reply:
Ultimately the presence of the `state` parameter tells us that it is initialized.